### PR TITLE
container-diff 0.6.2 (new formula)

### PR DESCRIPTION
**Pull request by sgerrand**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----
Happy to add more extensive tests, if required.

### CODE REVIEW

**Comment by ilovezfs** [comment]
>Happy to add more extensive tests, if required.

A functional test (i.e. not --help, not --version) is indeed required for new formulae.

File: Formula/container-diff.rb [hunk 1]
```diff
@@ -0,0 +1,34 @@
+class ContainerDiff < Formula
+  desc "Diff your Docker containers"
+  homepage "https://github.com/GoogleCloudPlatform/container-diff"
+  url "https://github.com/GoogleCloudPlatform/container-diff/archive/v0.5.1.tar.gz"
+  sha256 "2de7635f485314ad8506222783c615620a4a7c32b0d39215df70f1b8964c871f"
+
+  head do
```
Comment (ilovezfs) [inline Formula/container-diff.rb]: `head "https://github.com/GoogleCloudPlatform/container-diff.git"`

File: Formula/container-diff.rb [hunk 2]
```diff
@@ -0,0 +1,34 @@
+class ContainerDiff < Formula
+  desc "Diff your Docker containers"
+  homepage "https://github.com/GoogleCloudPlatform/container-diff"
+  url "https://github.com/GoogleCloudPlatform/container-diff/archive/v0.5.1.tar.gz"
+  sha256 "2de7635f485314ad8506222783c615620a4a7c32b0d39215df70f1b8964c871f"
+
```
Comment (ilovezfs) [inline Formula/container-diff.rb]: delete

File: Formula/container-diff.rb [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  head do
+    url "https://github.com/GoogleCloudPlatform/container-diff.git"
+  end
+
+  depends_on "go" => :build
+
+  def install
+    # The 0.5.1 release was tagged with an incorrect build version in the Makefile.
+    # It was fixed afterwards by https://github.com/GoogleCloudPlatform/container-diff/pull/142
```
Comment (ilovezfs) [inline Formula/container-diff.rb]: let's apply that as a patch instead to make sure this gets removed on the next update

File: Formula/container-diff.rb [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    # Until this is fixed, set the version manually.
+    ENV["VERSION_MAJOR"] = "0"
+    ENV["VERSION_MINOR"] = "5"
+    ENV["VERSION_BUILD"] = "1"
+
+    ENV["GOPATH"] = buildpath
+    mkdir_p buildpath/"src/github.com/GoogleCloudPlatform"
```
Comment (ilovezfs) [inline Formula/container-diff.rb]: mkpath

File: Formula/container-diff.rb [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    ln_sf buildpath, buildpath/"src/github.com/GoogleCloudPlatform/container-diff"
+
+    cd "src/github.com/GoogleCloudPlatform/container-diff" do
+      system "make"
+      bin.install ["out/container-diff"]
```
Comment (ilovezfs) [inline Formula/container-diff.rb]: [] is unneeded

**Review by ilovezfs — CHANGES_REQUESTED** [review]
looking forward to the improved test block :)

File: Formula/container-diff.rb (hunk 1 above)
Comment (DomT4) [reply Formula/container-diff.rb]: Going to add this to my _"I wonder if we can audit this"_ list locally. Single-line `head do` blocks have definitely made a comeback amongst contributors lately.

File: Formula/container-diff.rb (hunk 3 above)
Comment (DomT4) [reply Formula/container-diff.rb]: There's a new release upstream already, presumably with the fix in.

File: Formula/container-diff.rb (hunk 1 above)
Comment (ilovezfs) [reply Formula/container-diff.rb]: it's more than single-lined
```
head do
  url "foo.git",
      :branch => "bar"
end
```

Comment (ilovezfs) [reply Formula/container-diff.rb]: and `:using`

Comment (DomT4) [reply Formula/container-diff.rb]: Yeah, essentially anything with nothing but a `url` inside a `head do` block is what I meant, more specifically.

Comment (ilovezfs) [reply Formula/container-diff.rb]: I'll leave all that between you and rubocop. :see_no_evil:

File: Formula/container-diff.rb (hunk 3 above)
Comment (ilovezfs) [reply Formula/container-diff.rb]: even better

File: Formula/container-diff.rb (hunk 1 above)
Comment (DomT4) [reply Formula/container-diff.rb]: I loathe working with that thing 🙈. Great tool, dubious usability/syntax choices.

Comment (ilovezfs) [reply Formula/container-diff.rb]: yup. I'd rather chase contributors around about this than deal with that thing

File: Formula/container-diff.rb (hunk 3 above)
Comment (sgerrand) [reply Formula/container-diff.rb]: The maintainer [was receptive to feedback](https://github.com/GoogleCloudPlatform/container-diff/issues/147).

**Comment by sgerrand** [comment]
👋 @ilovezfs: I've implemented the changes you requested, including a functional test.

File: Formula/container-diff.rb [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+      -util-linux                        2.20.1-5.1ubuntu20.4                    1.6M
+      -vim-common                        2:7.4.052-1ubuntu3                      317K
+      -vim-tiny                          2:7.4.052-1ubuntu3                      931K
+      -whiptail                          0.52.15-2ubuntu5                        85K
+      -xauth                             1:1.0.7-1ubuntu1                        86K
+      -xkb-data                          2.10.1-1ubuntu1                         2.8M
+      -zlib1g                            1:1.2.8.dfsg-1ubuntu1                   171K
+    EOS
+
+    assert_match expected, shell_output("#{bin}/container-diff analyze docker/whalesay", 0)
```
Comment (DomT4) [inline Formula/container-diff.rb]: The `, 0` is redundant here. That's already the default check with `shell_output`.

File: Formula/container-diff.rb [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+      -udev                              204-5ubuntu20.11                        5M
+      -upstart                           1.12.1-0ubuntu4.2                       1.6M
+      -ureadahead                        0.100.0-16                              124K
+      -util-linux                        2.20.1-5.1ubuntu20.4                    1.6M
+      -vim-common                        2:7.4.052-1ubuntu3                      317K
+      -vim-tiny                          2:7.4.052-1ubuntu3                      931K
+      -whiptail                          0.52.15-2ubuntu5                        85K
+      -xauth                             1:1.0.7-1ubuntu1                        86K
+      -xkb-data                          2.10.1-1ubuntu1                         2.8M
+      -zlib1g                            1:1.2.8.dfsg-1ubuntu1                   171K
```
Comment (DomT4) [inline Formula/container-diff.rb]: Are some of these versions not likely to change over time, or?

**Comment by sgerrand** [comment]
I tried to use an image that did not change. Alternative suggestions are
welcome.

File: Formula/container-diff.rb (hunk 7 above)
Comment (sgerrand) [reply Formula/container-diff.rb]: I tried to use an image that is unlikely to change. Alternative suggestions are welcome.

**Comment by sgerrand** [comment]
Seeking a re-review. ♻️

**Comment by sgerrand** [comment]
> looking forward to the improved test block :)

👋 @ilovezfs

**Comment by sgerrand** [comment]
👋 @DomT4 @ilovezfs

File: Formula/container-diff.rb [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+  homepage "https://github.com/GoogleCloudPlatform/container-diff"
+  url "https://github.com/GoogleCloudPlatform/container-diff/archive/v0.5.2.tar.gz"
+  sha256 "9f64267b01e9f96efd053eadbac6435bdb15c05597e8ebe4e7081e13fb7e45cf"
+  head "https://github.com/GoogleCloudPlatform/container-diff.git"
+
+  depends_on "go" => :build
+
+  def install
+    ENV["GOPATH"] = buildpath
+    mkpath "src/github.com/GoogleCloudPlatform"
```
Comment (DomT4) [inline Formula/container-diff.rb]: `(buildpath/"src/github.com/GoogleCloudPlatform").mkpath`

Comment (sgerrand) [reply Formula/container-diff.rb]: 🙇  Added in 81f3661cbd. ♻️

**Comment by DomT4** [comment]
Thanks for the change! This is down to @ilovezfs now; it's been his yard for quite some time at this point.

**Comment by sgerrand** [comment]
@ilovezfs: Are there any other changes you'd like me to make?

**Comment by fxcoudert** [comment]
@BrewTestBot test this please

**Comment by ilovezfs** [comment]
@BrewTestBot test this please

**Comment by ilovezfs** [comment]
@BrewTestBot test this please

File: Formula/container-diff.rb (hunk 7 above)
Comment (ilovezfs) [reply Formula/container-diff.rb]: Can we get this data from a resource?

**Review by ilovezfs — CHANGES_REQUESTED** [review]

**Comment by kevmoo** [comment]
☹️

**Comment by sgerrand** [comment]
@ilovezfs: I'm unsure what changes if any are still required here. When you get a chance, would you please re-review this changeset.

**Comment by ilovezfs** [comment]
@sgerrand what I mean is, can we download the data that is in your big heredoc from someone instead of hard coding it into the formula, like

```
  resource "whalesay" do
    url "foo"
    sha256 "1234"
  end

  test do
    testpath.install resource("whalesay")
```

But I guess we'll just go with what you have for now.

**Review by ilovezfs — APPROVED** [review]

**Comment by ilovezfs** [comment]
Actually I just went for a smaller match `/gcc.*ubuntu/` which should be good enough for our purposes.

**Comment by ilovezfs** [comment]
@sgerrand The test doesn't seem to work for 0.6.2 anymore :(

**Review by ilovezfs — CHANGES_REQUESTED** [review]
we'll need the test working with the latest version of container-diff, which is 0.6.2 as of 23 Jan

**Comment by sgerrand** [comment]
Done!

**Review by ilovezfs — APPROVED** [review]

**Comment by ilovezfs** [comment]
Thanks for another contribution to Homebrew, @sgerrand! You're awesome! :white_check_mark:

**Pull request closed without merging**